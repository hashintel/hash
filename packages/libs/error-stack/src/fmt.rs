//! Implementation of display, which can be either fancy or "normal", depending
//! on the type color and non ascii character can be used.
//!
//! This is inspired by [miette](https://docs.rs/miette/latest/miette/index.html)

use alloc::{borrow::ToOwned, collections::VecDeque, format, string::String, vec, vec::Vec};
use core::fmt::{Debug, Display, Formatter, Write};

#[cfg(feature = "glyph")]
use owo_colors::{colored::Color, colors::Red, OwoColorize, Stream::Stdout};

use crate::{AttachmentKind, Frame, FrameKind, Report};

enum Instruction {
    Content(String),
    Entry { end: bool },
    Vertical,
    Indent,
}

impl Instruction {
    fn plain(text: String) -> Line {
        let mut queue = VecDeque::new();
        queue.push_back(Instruction::Content(text));

        queue
    }

    #[cfg(feature = "glyph")]
    fn with_red<F: FnOnce(&mut T) -> std::io::Result<()>, T: WriteColor>(
        writer: &mut T,
        f: F,
    ) -> std::io::Result<()> {
        let mut spec = ColorSpec::new();
        spec.set_fg(Some(Color::Red)).set_bold(true);

        writer.set_color(&spec)?;
        f(writer)?;
        writer.reset()
    }

    fn render(self, fmt: &mut Formatter) -> core::fmt::Result {
        match self {
            Instruction::Content(text) => fmt.write_str(&text),
            #[cfg(feature = "glyph")]
            Instruction::Entry { end: true } => {
                "╰─▶ ".if_supports_color(Stdout, |text| text.red()).fmt(fmt)
            }
            #[cfg(not(feature = "glyph"))]
            Instruction::Entry { end: true } => fmt.write_str(r#"\->"#),
            #[cfg(feature = "glyph")]
            Instruction::Entry { end: false } => {
                "├─▶ ".if_supports_color(Stdout, |text| text.red()).fmt(fmt)
            }
            #[cfg(not(feature = "glyph"))]
            Instruction::Entry { end: false } => fmt.write_str("|-> "),
            #[cfg(feature = "glyph")]
            Instruction::Vertical => "│   ".if_supports_color(Stdout, |text| text.red()).fmt(fmt),
            #[cfg(not(feature = "glyph"))]
            Instruction::Vertical => fmt.write_str("|   "),
            Instruction::Indent => fmt.write_str("    "),
        }
    }
}

type Line = VecDeque<Instruction>;
type Lines = Vec<Line>;

#[cfg(all(nightly, feature = "std"))]
fn backtrace<'a>(frame: &'a Frame, bt: &mut Vec<&'a std::backtrace::Backtrace>) -> Option<Line> {
    if let Some(backtrace) = frame.request_ref::<std::backtrace::Backtrace>() {
        bt.push(backtrace);

        Some(Instruction::plain(format!(
            "backtrace with {} frames ({})",
            backtrace.frames().len(),
            bt.len()
        )))
    } else {
        None
    }
}

#[cfg(feature = "spantrace")]
fn spantrace<'a>(frame: &'a Frame, st: &mut Vec<&'a tracing_error::SpanTrace>) -> Option<Line> {
    if let Some(span_trace) = frame.request_ref::<tracing_error::SpanTrace>() {
        let mut span = 0;
        span_trace.with_spans(|_, _| {
            span += 1;
            true
        });

        st.push(span_trace);

        Some(Instruction::plain(format!(
            "spantrace with {span} frames ({})",
            st.len()
        )))
    } else {
        None
    }
}

// we allow needless lifetime, as in some scenarios (where spantrace and backtrace are disabled)
// the lifetime is needless, but rewriting for that case would decrease readability.
#[allow(clippy::needless_lifetimes)]
#[allow(unused_variables)]
fn frame<'a>(
    frame: &'a Frame,
    defer: &mut Vec<Lines>,
    #[cfg(all(nightly, feature = "std"))] bt: &mut Vec<&'a std::backtrace::Backtrace>,
    #[cfg(feature = "spantrace")] st: &mut Vec<&'a tracing_error::SpanTrace>,
) -> Option<Lines> {
    // We allow `unused_mut` due to the fact that certain feature configurations will require
    // this to be mutable (backtrace and spantrace overwrite extend the lines)
    #[allow(unused_mut)]
    let mut lines = match frame.kind() {
        FrameKind::Context(context) => context
            .to_string()
            .split('\n')
            .map(ToOwned::to_owned)
            .collect(),
        FrameKind::Attachment(AttachmentKind::Opaque(_)) => vec![],
        FrameKind::Attachment(AttachmentKind::Printable(attachment)) => attachment
            .to_string()
            .split('\n')
            .map(ToOwned::to_owned)
            .collect(),
    };

    // defer collection of backtrace and spantrace after all messages, so that they are always last.
    #[cfg(all(nightly, feature = "std"))]
    if let Some(backtrace) = backtrace(frame, bt) {
        defer.push(vec![backtrace]);

        return None;
    }

    #[cfg(feature = "spantrace")]
    if let Some(spantrace) = spantrace(frame, st) {
        defer.push(vec![spantrace]);

        return None;
    }

    Some(lines.into_iter().map(Instruction::plain).collect())
}

// we allow needless lifetime, as in some scenarios (where spantrace and backtrace are disabled)
// the lifetime is needless, but rewriting for that case would decrease readability.
#[allow(clippy::needless_lifetimes)]
fn frame_root<'a>(
    root: &'a Frame,
    #[cfg(all(nightly, feature = "std"))] bt: &mut Vec<&'a std::backtrace::Backtrace>,
    #[cfg(feature = "spantrace")] st: &mut Vec<&'a tracing_error::SpanTrace>,
) -> Lines {
    let mut plain = vec![root];

    let next;
    let mut ptr = root;

    // find all the frames that are part of this stack,
    // meaning collect them until we hit the end or a group of multiple frames.
    loop {
        let sources = ptr.sources();

        next = match sources {
            [] => None,
            [child] => {
                plain.push(child);
                ptr = child;
                continue;
            }
            group => Some(group),
        };

        break;
    }

    let mut groups = vec![];
    let mut defer = vec![];

    let mut opaque = 0;
    for child in plain {
        let content = frame(
            child,
            &mut defer,
            #[cfg(all(nightly, feature = "std"))]
            bt,
            #[cfg(feature = "spantrace")]
            st,
        );

        if let Some(lines) = content {
            if lines.is_empty() {
                opaque += 1;
            } else {
                groups.push(lines);
            }
        }
    }

    groups.append(&mut defer);

    match opaque {
        0 => {}
        1 => {
            groups.push(vec![Instruction::plain(
                "1 additional attachment".to_owned(),
            )]);
        }
        n => {
            groups.push(vec![Instruction::plain(format!(
                "{n} additional attachments"
            ))]);
        }
    }

    if let Some(group) = next {
        for child in group {
            let content = frame_root(
                child,
                #[cfg(all(nightly, feature = "std"))]
                bt,
                #[cfg(feature = "spantrace")]
                st,
            );

            if !content.is_empty() {
                groups.push(content);
            }
        }
    }

    // The first item is always the title,
    // after that every group gets one of the following indents: `|->`, or `\->`
    // if it is the last one.
    let total = groups.len();
    groups
        .into_iter()
        .enumerate()
        .flat_map(|(pos, content)| {
            let last = pos == total - 1;
            let first = pos == 0;

            content.into_iter().enumerate().map(move |(idx, mut line)| {
                if first {
                    // the first line is the title, therefore not indented.
                } else if last {
                    if idx == 0 {
                        line.push_front(Instruction::Entry { end: true });
                    } else {
                        line.push_front(Instruction::Indent);
                    }
                } else if idx == 0 {
                    line.push_front(Instruction::Entry { end: false });
                } else {
                    line.push_front(Instruction::Vertical);
                }

                line
            })
        })
        .collect()
}

pub fn report<C>(report: &Report<C>) -> String {
    #[cfg(all(nightly, feature = "std"))]
    let mut bt = Vec::new();
    #[cfg(feature = "spantrace")]
    let mut st = Vec::new();

    let mut lines = vec![];
    for frame in report.current_frames() {
        let display = frame_root(
            frame,
            #[cfg(all(nightly, feature = "std"))]
            &mut bt,
            #[cfg(feature = "spantrace")]
            &mut st,
        );

        lines.extend(display);
        lines.push(Instruction::plain("".to_owned()));
    }

    #[cfg(all(nightly, feature = "std"))]
    {
        lines.extend(vec!["".to_owned(); 2]);

        for (pos, backtrace) in bt.into_iter().enumerate() {
            lines.push(Instruction::plain(format!("Backtrace No. {}", pos + 1)));
            lines.extend(
                backtrace
                    .to_string()
                    .split('\n')
                    .map(ToOwned::to_owned)
                    .map(Instruction::plain),
            );
        }
    }

    #[cfg(feature = "spantrace")]
    {
        lines.extend(vec!["".to_owned(); 2]);

        for (pos, span_trace) in st.into_iter().enumerate() {
            lines.push(Instruction::plain(format!("Spantrace No. {}", pos + 1)));
            lines.extend(
                span_trace
                    .to_string()
                    .split('\n')
                    .map(ToOwned::to_owned)
                    .map(Instruction::plain),
            );
        }
    }

    lines.into_iter().fold(String::new(), |mut acc, line| {
        acc.push('\n');
        acc.push_str(&line);
        acc
    })
}

mod hook {
    use std::{
        any::{Any, Demand, Provider, TypeId},
        collections::HashMap,
        marker::PhantomData,
    };

    use crate::Frame;

    struct AnyContext {
        defer: Vec<Vec<String>>,
        inner: HashMap<TypeId, Box<dyn Any>>,
    }

    struct Context<'a, T> {
        parent: &'a mut AnyContext,
        _marker: PhantomData<T>,
    }

    impl<T> Context<T> {
        pub fn defer(&mut self, value: String) {
            self.0.push(value.split('\n').collect());
        }

        pub fn defer_lines<L: IntoIterator<Item = String>>(&mut self, lines: L) {
            self.0.push(lines.into_iter().collect())
        }
    }

    impl<T: 'static> Context<T> {
        fn get<U>(&self) -> Option<&U> {
            let id = T::type_id();

            let inner = self.parent.inner.get(&id)?;
            inner.downcast_ref()
        }

        fn get_mut<U>(&mut self) -> Option<&mut U> {
            let id = T::type_id();

            let inner = self.parent.inner.get_mut(&id)?;
            inner.downcast_mut()
        }

        fn insert<U>(&mut self, value: U) -> Option<&U> {
            let id = T::type_id();

            let inner = self.parent.inner.insert(id, Box::new(value));
            let inner = inner.as_ref()?;
            inner.downcast_ref()
        }
    }

    pub trait Hook<T> {
        fn call(&self, frame: &T, defer: &mut AnyContext) -> Option<String>;
    }

    impl<F, T> Hook<T> for F
    where
        F: Fn(&T, &mut Context<T>) -> String,
        T: Send + Sync + 'static,
    {
        fn call(&self, frame: &T, defer: &mut AnyContext) -> Option<String> {
            Some((self)(frame, defer))
        }
    }

    pub struct Stack<L, T, R> {
        left: L,
        right: R,
        _marker: PhantomData<T>,
    }

    impl<L, T, R> Hook<Frame> for Stack<L, T, R>
    where
        L: Hook<T>,
        T: Send + Sync + 'static,
        R: Hook<Frame>,
    {
        fn call(&self, frame: &Frame, defer: &mut AnyContext) -> Option<String> {
            if let Some(frame) = frame.downcast_ref::<T>() {
                self.left.call(frame, defer)
            } else {
                self.right.call(frame, defer)
            }
        }
    }

    impl<T> Hook<T> for () {
        fn call(&self, _: &T, _: &mut AnyContext) -> Option<String> {
            None
        }
    }

    pub struct Both<L, R> {
        left: L,
        right: R,
    }

    impl<L, R> Hook<Frame> for Both<L, R>
    where
        L: Hook<Frame>,
        R: Hook<Frame>,
    {
        fn call(&self, frame: &Frame, defer: &mut AnyContext) -> Option<String> {
            self.left
                .call(frame, defer)
                .or_else(|| self.right.call(frame, defer))
        }
    }

    impl Hook<Frame> for Box<dyn Hook<Frame>> {
        fn call(&self, frame: &Frame, defer: &mut AnyContext) -> Option<String> {
            let hook = self.as_ref();

            hook.call(frame, defer)
        }
    }

    struct Hooks<T: Hook<Frame>>(T);

    impl Hooks<()> {
        pub fn new() -> Self {
            Self(())
        }
    }

    impl<T: Hook<Frame>> Hooks<T> {
        fn new_with(hook: T) -> Self {
            Hooks(hook)
        }

        pub fn push<U: Hook<V>, V: Send + Sync + 'static>(self, hook: T) -> Hooks<Stack<U, V, T>> {
            let stack = Stack {
                left: hook,
                right: self.0,
                _marker: PhantomData::default(),
            };

            Hooks::new_with(stack)
        }

        pub fn append<U: Hook<Frame>>(self, other: Hooks<U>) -> Hooks<Both<T, U>> {
            let both = Both {
                left: self.0,
                right: other.0,
            };

            Hooks::new_with(both)
        }
    }

    impl<T: Hook<Frame> + 'static> Hooks<T> {
        pub fn erase(self) -> ErasedHooks {
            Hooks::new_with(Box::new(self.0))
        }
    }

    type ErasedHooks = Hooks<Box<dyn Hook<Frame>>>;

    mod builtin {
        #[cfg(all(nightly, feature = "std"))]
        use std::backtrace::Backtrace;

        #[cfg(feature = "spantrace")]
        use tracing_error::SpanTrace;

        use crate::fmt::hook::{AnyContext, Context};

        #[cfg(all(nightly, feature = "std"))]
        fn backtrace(backtrace: &Backtrace, ctx: &mut Context<Backtrace>) -> String {
            let idx = match ctx.get::<usize>().copied() {
                None => {
                    ctx.insert(0);
                    0
                }
                Some(idx) => idx,
            };

            ctx.defer(format!("Backtrace No. {}\n{}", idx + 1, backtrace));
            ctx.insert(idx + 1);

            format!(
                "backtrace with {} frames ({})",
                backtrace.frames().len(),
                idx + 1
            )
        }

        #[cfg(feature = "spantrace")]
        fn spantrace(spantrace: &SpanTrace, ctx: &mut Context<SpanTrace>) -> String {
            let idx = match ctx.get::<usize>().copied() {
                None => {
                    ctx.insert(0);
                    0
                }
                Some(idx) => idx,
            };

            let mut span = 0;
            spantrace.with_spans(|_, _| {
                span += 1;
                true
            });

            ctx.defer(format!("Span Trace No. {}\n{}", idx + 1, spantrace));
            ctx.insert(idx + 1);

            format!("spantrace with {span} frames ({})", idx + 1)
        }
    }
}
