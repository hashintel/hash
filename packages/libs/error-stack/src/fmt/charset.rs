use core::sync::atomic::{AtomicU8, Ordering};

use crate::{
    fmt::r#override::{AtomicOverride, AtomicPreference},
    Report,
};

/// The available supported charsets
///
/// Can be accessed through [`crate::fmt::HookContext::charset`], and set via
/// [`Report::set_charset`].
#[derive(Debug, Copy, Clone)]
pub enum Charset {
    /// Terminal of the user supports utf-8
    ///
    /// This is the default if no charset has been explicitly set.
    Utf8,

    /// Terminal of the user supports ASCII
    Ascii,
}

impl Charset {
    pub(super) fn load() -> Self {
        // we assume that most fonts and terminals nowadays support Utf8, which is why this is
        // the default
        CHARSET_OVERRIDE
            .load()
            .map_or(Self::Utf8, |charset| charset)
    }
}

/// Value layout:
/// `0x00`: `Charset::Ascii`
/// `0x01`: `Charset::Utf8`
///
/// all others: unset/none
impl AtomicPreference for Charset {
    fn load(value: u8) -> Option<Self> {
        match value {
            0x00 => Some(Self::Ascii),
            0x01 => Some(Self::Utf8),
            _ => None,
        }
    }

    fn store(value: Option<Self>) -> u8 {
        match value {
            None => 0xFF,
            Some(Self::Ascii) => 0x00,
            Some(Self::Utf8) => 0x01,
        }
    }
}

static CHARSET_OVERRIDE: AtomicOverride<Charset> = AtomicOverride::new();

impl Report<()> {
    /// Set the charset preference
    ///
    /// If the value is [`None`], a previously set preference will be unset, while with [`Some`] a
    /// specific charset will be set.
    ///
    /// The value defaults to [`Charset::Utf8`].
    ///
    /// # Example
    ///
    /// ```
    /// # // we only test the snapshot on rust 1.65, therefore report is unused (so is render)
    /// # #![cfg_attr(not(rust_1_65), allow(dead_code, unused_variables, unused_imports))]
    /// use std::io::{Error, ErrorKind};
    ///
    /// use error_stack::{report, Report};
    /// use error_stack::fmt::{Charset};
    ///
    /// struct Suggestion(&'static str);
    ///
    /// Report::install_debug_hook::<Suggestion>(|Suggestion(value), context| {
    ///     match context.charset() {
    ///         Charset::Utf8 => context.push_body(format!("📝 {value}")),
    ///         Charset::Ascii => context.push_body(format!("suggestion: {value}"))
    ///     };
    /// });
    ///
    /// let report =
    ///     report!(Error::from(ErrorKind::InvalidInput)).attach(Suggestion("oh no, try again"));
    ///
    /// # fn render(value: String) -> String {
    /// #     let backtrace = regex::Regex::new(r"backtrace no\. (\d+)\n(?:  .*\n)*  .*").unwrap();
    /// #     let backtrace_info = regex::Regex::new(r"backtrace( with (\d+) frames)? \((\d+)\)").unwrap();
    /// #
    /// #     let value = backtrace.replace_all(&value, "backtrace no. $1\n  [redacted]");
    /// #     let value = backtrace_info.replace_all(value.as_ref(), "backtrace ($3)");
    /// #
    /// #     ansi_to_html::convert_escaped(value.as_ref()).unwrap()
    /// # }
    /// #
    /// Report::set_charset(Some(Charset::Utf8));
    /// println!("{report:?}");
    /// # #[cfg(rust_1_65)]
    /// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__charset_utf8.snap")].assert_eq(&render(format!("{report:?}")));
    ///
    /// Report::set_charset(Some(Charset::Ascii));
    /// println!("{report:?}");
    /// # #[cfg(rust_1_65)]
    /// # expect_test::expect_file![concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__charset_ascii.snap")].assert_eq(&render(format!("{report:?}")));
    /// ```
    ///
    /// Which will result in something like:
    ///
    /// <pre>
    #[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__charset_utf8.snap"))]
    /// </pre>
    ///
    /// <pre>
    #[doc = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/snapshots/doc/fmt__charset_ascii.snap"))]
    /// </pre>
    pub fn set_charset(charset: Option<Charset>) {
        CHARSET_OVERRIDE.store(charset);
    }
}
