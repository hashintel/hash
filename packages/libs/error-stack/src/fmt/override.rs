use core::{
    marker::PhantomData,
    sync::atomic::{AtomicU8, Ordering},
};

pub(super) trait AtomicPreference {
    const DEFAULT: Self;

    fn from_u8(value: u8) -> Self;
    fn into_u8(self) -> u8;
}

pub(super) struct AtomicOverride<T: AtomicPreference> {
    inner: AtomicU8,
    _marker: PhantomData<fn() -> *const T>,
}

impl<T: AtomicPreference> AtomicOverride<T> {
    pub(super) const fn new() -> Self {
        Self {
            inner: AtomicU8::new(T::DEFAULT.store()),
            _marker: PhantomData,
        }
    }

    pub(super) fn store(&self, value: T) {
        self.inner.store(value.into_u8(), Ordering::Relaxed);
    }

    pub(super) fn load(&self) -> T {
        let inner = self.inner.load(Ordering::Relaxed);

        T::from_u8(inner)
    }
}
