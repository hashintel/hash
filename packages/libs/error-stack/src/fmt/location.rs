use core::{
    fmt,
    fmt::{Display, Formatter},
    panic::Location,
};

use crate::fmt::{
    color::{Color, DisplayStyle, Style},
    ColorMode,
};

pub(super) struct LocationDisplay<'a> {
    location: &'a Location<'static>,
    mode: ColorMode,
}

impl<'a> LocationDisplay<'a> {
    #[must_use]
    pub(super) const fn new(location: &'a Location<'static>, mode: ColorMode) -> Self {
        Self { location, mode }
    }
}

impl<'a> Display for LocationDisplay<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let location = self.location;

        let mut style = Style::new();

        match self.mode {
            ColorMode::None => {}
            ColorMode::Color => style.set_foreground(Color::Black, true),
            ColorMode::Emphasis => style.set_display(DisplayStyle::new().with_italic(true)),
        };

        f.write_fmt(format_args!("at {}", style.apply(&location)))?;

        Ok(())
    }
}
