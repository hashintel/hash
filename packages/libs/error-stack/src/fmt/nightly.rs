use crate::fmt::Line;

pub struct DebugDiagnostic {
    output: Line,
    text: Vec<String>,
}

impl DebugDiagnostic {
    pub fn next(output: String) -> Self {
        Self {
            output: Line::Next(output),
            text: vec![],
        }
    }

    pub fn defer(output: String) -> Self {
        Self {
            output: Line::Defer(output),
            text: vec![],
        }
    }

    pub fn and_text(mut self, text: String) -> Self {
        self.text.push(text);
        self
    }

    pub(crate) fn output(&self) -> &Line {
        &self.output
    }

    pub(crate) fn text(&self) -> &[String] {
        &self.text
    }
}
