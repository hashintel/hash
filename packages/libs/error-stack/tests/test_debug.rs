#![cfg_attr(nightly, feature(provide_any))]
#![cfg_attr(all(nightly, feature = "std"), feature(backtrace))]

mod common;
use common::*;
#[cfg(all(nightly, feature = "experimental"))]
use error_stack::fmt::DebugDiagnostic;
#[cfg(feature = "hooks")]
use error_stack::fmt::{HookContext, Hooks, Line};
#[allow(unused_imports)]
use error_stack::Report;
use insta::assert_snapshot;
#[cfg(feature = "glyph")]
use owo_colors::set_override;
#[cfg(feature = "spantrace")]
use tracing_error::ErrorLayer;
#[cfg(feature = "spantrace")]
use tracing_subscriber::layer::SubscriberExt;

#[cfg(feature = "spantrace")]
fn install_tracing_subscriber() {
    static ONCE: std::sync::Once = std::sync::Once::new();
    ONCE.call_once(|| {
        tracing::subscriber::set_global_default(
            tracing_subscriber::Registry::default().with(ErrorLayer::default()),
        )
        .expect("Could not set tracing subscriber");
    })
}

#[cfg(not(feature = "spantrace"))]
fn install_tracing_subscriber() {}

#[cfg(not(all(nightly, feature = "std")))]
fn install_backtrace() {
    std::env::set_var("RUST_LIB_BACKTRACE", "0");
}

#[cfg(all(nightly, feature = "std"))]
fn install_backtrace() {
    std::env::set_var("RUST_LIB_BACKTRACE", "1");
}

#[cfg(feature = "glyph")]
fn force_color() {
    set_override(false);
}

#[cfg(not(feature = "glyph"))]
fn force_color() {}

fn snap_suffix() -> String {
    install_tracing_subscriber();
    install_backtrace();

    #[allow(unused_mut)]
    let mut suffix: Vec<&'static str> = vec![];

    #[cfg(feature = "spantrace")]
    {
        suffix.push("spantrace");
    }

    #[cfg(all(nightly, feature = "std"))]
    if supports_backtrace() {
        suffix.push("backtrace");
    }

    #[cfg(feature = "glyph")]
    {
        suffix.push("glyph");
    }

    force_color();

    suffix.join("-")
}

/// Overwrite to create spantraces when necessary.
#[cfg(feature = "spantrace")]
pub fn create_report() -> Report<RootError> {
    #[tracing::instrument]
    fn func_b() -> error_stack::Result<(), RootError> {
        create_error()
    }

    #[tracing::instrument]
    fn func_a() -> error_stack::Result<(), RootError> {
        func_b()
    }

    capture_error(func_a)
}

/// This is taken from the rstest pattern https://insta.rs/docs/patterns/
macro_rules! set_snapshot_suffix {
    () => {{
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(snap_suffix());
        #[allow(deprecated)]
        settings.bind_to_thread();
    }};
}

/// The provider API extension via `DebugDiagnostic` is only available under experimental and
/// nightly
#[test]
#[cfg(all(nightly, feature = "experimental"))]
fn provider() {
    set_snapshot_suffix!();

    let mut report = create_report().attach_printable(PrintableA(0));
    report.extend_one({
        let mut report = create_report().attach_printable(PrintableB(1));

        report.extend_one(
            create_report()
                .attach(DebugDiagnostic::next("ABC".to_owned()))
                .attach(AttachmentA(1))
                .attach_printable(PrintableB(1)),
        );

        report.attach(AttachmentA(2)).attach_printable("Test")
    });

    assert_snapshot!(format!("{report:?}"));
}

/// The provider API extension via `DebugDiagnostic` is only available under experimental and
/// nightly
#[test]
#[cfg(all(nightly, feature = "experimental"))]
fn provider_ext() {
    set_snapshot_suffix!();

    let mut report = create_report().attach_printable(PrintableA(0));
    report.extend_one({
        let mut report = create_report().attach_printable(PrintableB(1));

        report.extend_one(
            create_report()
                .attach(DebugDiagnostic::next("ABC".to_owned()))
                .attach(AttachmentA(1))
                .attach_printable(PrintableB(1)),
        );

        report.attach(AttachmentA(2)).attach_printable("Test")
    });

    assert_snapshot!(format!("{report:#?}"));
}

#[test]
fn linear() {
    set_snapshot_suffix!();

    let report = create_report()
        .attach_printable(PrintableA(0))
        .attach(AttachmentA)
        .attach(AttachmentB)
        .change_context(ContextA(0))
        .attach_printable(PrintableB(0))
        .attach(AttachmentB)
        .change_context(ContextB(0))
        .attach_printable("Printable C");

    assert_snapshot!(format!("{report:?}"));
}

#[test]
fn linear_ext() {
    set_snapshot_suffix!();

    let report = create_report()
        .attach_printable(PrintableA(0))
        .attach(AttachmentA)
        .attach(AttachmentB)
        .change_context(ContextA(0))
        .attach_printable(PrintableB(0))
        .attach(AttachmentB)
        .change_context(ContextB(0))
        .attach_printable("Printable C");

    assert_snapshot!(format!("{report:#?}"));
}

#[test]
fn complex() {
    set_snapshot_suffix!();

    let mut report = create_report().attach_printable(PrintableA(0));
    report.extend_one({
        let mut report = create_report().attach_printable(PrintableB(1));

        report.extend_one(
            create_report()
                .attach(AttachmentB(0))
                .attach(AttachmentA(1))
                .attach_printable(PrintableB(1)),
        );

        report.attach(AttachmentA(2)).attach_printable("Test")
    });

    assert_snapshot!(format!("{report:?}"));
}

#[test]
fn location_edge_case() {
    set_snapshot_suffix!();

    let report = create_report();

    assert_snapshot!(format!("{report:?}"));
}

#[cfg(feature = "hooks")]
mod hooks {
    //! To be able to set `OnceCell` multiple times we need to run each of them in a separate fork,
    //! this works without fault in `nextest` you cannot guarantee that this is always the case.
    //! This is mostly undocumented behavior.

    use super::*;

    #[test]
    fn hook() {
        set_snapshot_suffix!();

        let report = create_report().attach(2u32);

        Report::install_hook(Hooks::new().push(|_: &u32| Line::next("unsigned 32bit integer")))
            .unwrap();

        assert_snapshot!("hook", format!("{report:?}"));
    }

    #[test]
    fn hook_context() {
        set_snapshot_suffix!();

        let report = create_report().attach(2u32);

        Report::install_hook(Hooks::new().push(|_: &u32, ctx: &mut HookContext<u32>| {
            Line::next(format!("unsigned 32bit integer (No. {})", ctx.incr()))
        }))
        .unwrap();

        assert_snapshot!("hook_context", format!("{report:?}"));
    }

    #[test]
    fn hook_stack() {
        set_snapshot_suffix!();

        let report = create_report().attach(1u32).attach(2u64);

        Report::install_hook(
            Hooks::new()
                .push(|_: &u32| Line::next("unsigned 32bit integer"))
                .push(|_: &u64| Line::next("unsigned 64-bit integer")),
        )
        .unwrap();

        assert_snapshot!("hook_stack", format!("{report:?}"));
    }

    #[test]
    fn hook_combine() {
        set_snapshot_suffix!();

        let report = create_report() //
            .attach(1u32)
            .attach(2u64)
            .attach(3u16);

        let other = Hooks::new()
            .push(|_: &u32| Line::next("u32 (other)"))
            .push(|_: &u16| Line::next("u16 (other)"));

        Report::install_hook(
            Hooks::new()
                .push(|_: &u32| Line::next("u32"))
                .push(|_: &u64| Line::next("u64"))
                .combine(other),
        )
        .unwrap();

        assert_snapshot!("hook_combine", format!("{report:?}"));
    }

    #[test]
    fn hook_defer() {
        set_snapshot_suffix!();

        let report = create_report() //
            .attach(1u32)
            .attach(2u64)
            .attach(3u16);

        Report::install_hook(
            Hooks::new()
                .push(|_: &u32| Line::defer("u32"))
                .push(|_: &u64| Line::next("u64"))
                .push(|_: &u16| Line::defer("u16")),
        )
        .unwrap();

        assert_snapshot!("hook_defer", format!("{report:?}"));
    }

    #[test]
    fn hook_decr() {
        set_snapshot_suffix!();

        let report = create_report() //
            .attach(1u32)
            .attach(2u32)
            .attach(3u32);

        Report::install_hook(
            Hooks::new()
                .push(|_: &u32, ctx: &mut HookContext<u32>| Line::next(format!("{}", ctx.decr()))),
        )
        .unwrap();

        assert_snapshot!("hook_decr", format!("{report:?}"));
    }

    #[test]
    fn hook_incr() {
        set_snapshot_suffix!();

        let report = create_report() //
            .attach(1u32)
            .attach(2u32)
            .attach(3u32);

        Report::install_hook(
            Hooks::new()
                .push(|_: &u32, ctx: &mut HookContext<u32>| Line::next(format!("{}", ctx.incr()))),
        )
        .unwrap();

        assert_snapshot!("hook_incr", format!("{report:?}"));
    }
}
