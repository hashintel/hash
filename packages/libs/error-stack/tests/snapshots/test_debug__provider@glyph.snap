---
source: tests/test_debug.rs
expression: "format!(\"{report:?}\")"
---

Printable A
â”‚ ğŸ“Œ tests/test_debug.rs:112:38
â•°â”€â–¶ Root error
    â•° ğŸ“Œ tests/common.rs:140:5

Test
â”‚ ğŸ“Œ tests/test_debug.rs:123:39
â”œâ”€â–¶ 1 additional attachment
â”œâ”€â–¶ Printable B
â”‚   â”‚ ğŸ“Œ tests/test_debug.rs:114:42
â”‚   â•°â”€â–¶ Root error
â”‚       â•° ğŸ“Œ tests/common.rs:140:5
â•°â”€â–¶ Printable B
    â”‚ ğŸ“Œ tests/test_debug.rs:120:18
    â”œâ”€â–¶ ABC
    â”‚   â•° ğŸ“Œ tests/test_debug.rs:118:18
    â”œâ”€â–¶ Root error
    â”‚   â•° ğŸ“Œ tests/common.rs:140:5
    â•°â”€â–¶ 1 additional attachment

